### ISSUE TITLE
JEP 328 Flight Recorder

### ISSUE THREAD

**Issue by pshipton** [eclipse-openj9/openj9#1876]
https://bugs.openjdk.java.net/browse/JDK-8193393

This is an "Epic" to contain requirements for supporting this JEP on OpenJ9.

**Comment by pshipton** [comment]
I don't believe there are any plans to implement this.
@DanHeidinga

**Comment by fjeremic** [comment]
Isn't this something we are still looking into? What is the alternative for OpenJ9? HealthCenter?

**Comment by DanHeidinga** [comment]
> Isn't this something we are still looking into? What is the alternative for OpenJ9? HealthCenter?

Currently, Healthcenter is the alternative for OpenJ9.  This is an area we're looking more broadly to determine the right way to get metrics out of the JVM.  FlightRecorder is one option but it's not the only one.

Closing this makes sense as this item is unlikely to be the best place to track OpenJ9's final approach to this area

**Comment by MariusVolkhart** [comment]
@DanHeidinga is there another public location where one can follow along or join the conversation about OpenJ9's approach?

I'm primarily interested in how to do custom events on OpenJ9 (`jdk.jfr.*` and `jdk.jfr.consumer.*` equivalents). I looked through various [Eclipse](https://www.eclipse.org/openj9/docs/diag_overview/) and [IBM](https://www.ibm.com/docs/en/mon-diag-tools?topic=center-using-health-api) documentation but didn't see this functionality pop up.

**Comment by danjhop3** [comment]
IT IS NEEDED AT A JAR LEVEL

**Comment by DanHeidinga** [comment]
> I'm primarily interested in how to do custom events on OpenJ9 (`jdk.jfr.*` and `jdk.jfr.consumer.*` equivalents). I looked through various [Eclipse](https://www.eclipse.org/openj9/docs/diag_overview/) and [IBM](https://www.ibm.com/docs/en/mon-diag-tools?topic=center-using-health-api) documentation but didn't see this functionality pop up.

Does the Trace Java API [1] provide the required functionality?  It allows the application to register tracepoints and inject them into the per-thread tracebuffers.  These are - by default - an in-memory ring buffer that records the last few events or can be redirected to a file.

> is there another public location where one can follow along or join the conversation about OpenJ9's approach?

This is a really good question.  IBM HealthCenter is owned by IBM and to my knowledge, hasn't been opened sourced.  The OpenJ9 project had relied on HealthCenter for historical reasons (and because HC supported OpenJ9 and was available). I'm not sure what the current state of HC is and what kind of resources are behind it today.  Maybe @mstoodle or @pshipton can comment?

This may be an area we need to re-investigate if the current path still works or if we should give JFG / Mission control another look

**Comment by mstoodle** [comment]
You're right, it hasn't been open sourced, although nothing to my knowledge would stop us from doing it (other than the fact that it would probably be pretty close to a dump and run :( ). There are a few people who help out with support for HC, but no dedicated resourcing right now.

I would support reopening the discussion to look at JFR / Mission Control, but unfortunately can't commit people to help with such an effort. This is an area where we could really use some help. I'd like to believe the existing hook/trace framework represents some significant fraction of the work that would be needed if we were to go down that path (so we wouldn't be starting from nothing: many of the needed hook points are probably already there), but that doesn't mean there wouldn't still be a fair amount of work needed :( .

**Comment by Wood-Xia** [comment]
I also strongly support to reopen this discussion.
When we switch from OpenJDK to OpenJ9, we are happy with its low memory footprint.
Well, when we are looking for some profiling tool as JFR which is cool, we are  a little disappointed....

**Comment by monwolf** [comment]
We are at the same point, we are really happy with the low memory footprint but we lost a good capacity of our current APM system that is the real-time profiling of the APP.

**Comment by pshipton** [comment]
Re-opening since there is active development on JFR for OpenJ9 now.

**Comment by andrewazores** [comment]
I see that the latest OpenJ9 releases have support for JFR jcmds, ex. `jcmd 1 JFR.dump` - nice progress! Is it also a goal to implement the [FlightRecorder internal API](https://docs.oracle.com/en/java/javase/11/docs/api/jdk.jfr/jdk/jfr/FlightRecorder.html)?

Edit to add: also, the FlightRecorderMXBean?

**Comment by pshipton** [comment]
@tajila pls see the question in the previous comment.

**Comment by tajila** [comment]
@andrewazores Yes, we plan to add more capabilities in the upcoming releases. I anticipate we may have support for the jdk.jfr.* APIs and FlightRecorderMXBean in 1H next year.

**Comment by andrewazores** [comment]
Perfect, thanks for the info @tajila .
